### PR TITLE
Issue/#49 Write integration tests

### LINKED ISSUE #49: Write integration tests
Currently, testing relies on unit tests with mocked dependencies. Time to test the real integrations in the app - write integration tests!

### PR DESCRIPTION
**Pull request by devaajanne**
Closes #49 

This pull request introduces several improvements to error handling, integration testing, and dependency management in the backend service. The changes enhance the robustness of exception handling by introducing a global exception handler, improve integration test coverage for key services, and update dependencies to support these improvements and enable containerized database testing.

**Error handling improvements:**
- Added a global exception handler (`GlobalExceptionHandler`) to provide consistent HTTP error responses for `IllegalArgumentException` (400 Bad Request) and `IllegalStateException` (503 Service Unavailable) across the application.
- Updated `RepositoryService` to throw `IllegalArgumentException` instead of generic `RuntimeException` for unknown language cases and documented possible exceptions in Javadoc comments for relevant methods. [[1]](diffhunk://#diff-b1b7da461414f71975e343c4860d46bf9081ae1d0d8d6e5530fe551f8c946bdfR32) [[2]](diffhunk://#diff-b1b7da461414f71975e343c4860d46bf9081ae1d0d8d6e5530fe551f8c946bdfL43-R44) [[3]](diffhunk://#diff-b1b7da461414f71975e343c4860d46bf9081ae1d0d8d6e5530fe551f8c946bdfR64) [[4]](diffhunk://#diff-b1b7da461414f71975e343c4860d46bf9081ae1d0d8d6e5530fe551f8c946bdfL74-R76) [[5]](diffhunk://#diff-b1b7da461414f71975e343c4860d46bf9081ae1d0d8d6e5530fe551f8c946bdfR96) [[6]](diffhunk://#diff-b1b7da461414f71975e343c4860d46bf9081ae1d0d8d6e5530fe551f8c946bdfL107-R110)

**Testing and integration:**
- Added comprehensive integration tests for `GameService` and `RepositoryService`, covering normal and edge cases, including error scenarios and database state changes. [[1]](diffhunk://#diff-de38318949bf44885b5b577731e5c2dfdca5fcd5c7098bdd73d1fffad8604804R1-R230) [[2]](diffhunk://#diff-dcb965443fc72e702b561ef3142904945f0b09f8ff4ee44255a7bc0c137081cdR1-R138)
- Enhanced the main application test (`BackendApplicationTests`) to use Testcontainers for PostgreSQL, ensuring reliable integration testing with an isolated database environment.
- Removed the obsolete `PlaceholderIntegrationTest` file.

**Dependency management:**
- Replaced `spring-boot-starter-web` with `spring-boot-starter-webmvc` and added dependencies for `spring-boot-starter-webmvc-test` and Testcontainers (JUnit Jupiter and PostgreSQL) to support advanced integration testing.

### CODE REVIEW

File: backend/src/test/java/backend/integration/GameControllerIntegrationTests.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+@AutoConfigureMockMvc
+@Testcontainers
+@Tag("integrationTest")
+class GameControllerIntegrationTest {
+  @Autowired private MockMvc mockMvc;
+
+  @Container static PostgreSQLContainer postgres = new PostgreSQLContainer("postgres:18.1");
+
+  @LocalServerPort private Integer port;
```
[the remark this answers is not in the document]
Comment (devaajanne) [reply backend/src/test/java/backend/integration/GameControllerIntegrationTests.java]: Fixed

File: backend/build.gradle [hunk 2]
```diff
@@ -31,12 +31,15 @@ dependencies {
...
 	implementation 'org.springframework.boot:spring-boot-starter-data-jpa'
 	implementation 'org.springframework.boot:spring-boot-starter-validation'
-	implementation 'org.springframework.boot:spring-boot-starter-web'
+	implementation 'org.springframework.boot:spring-boot-starter-webmvc'
 	implementation 'org.springframework.boot:spring-boot-starter-security'
 	compileOnly 'org.projectlombok:lombok'
 	runtimeOnly 'org.postgresql:postgresql'
 	annotationProcessor 'org.projectlombok:lombok'
 	testImplementation 'org.springframework.boot:spring-boot-starter-test'
+	testImplementation 'org.springframework.boot:spring-boot-starter-webmvc-test'
```
[the remark this answers is not in the document]
Comment (devaajanne) [reply backend/build.gradle:40]: It is actually needed

File: backend/src/test/java/backend/integration/GameControllerIntegrationTests.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+import org.springframework.test.web.servlet.MockMvc;
+import org.testcontainers.junit.jupiter.Container;
+import org.testcontainers.junit.jupiter.Testcontainers;
+import org.testcontainers.postgresql.PostgreSQLContainer;
+
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+@AutoConfigureMockMvc
+@Testcontainers
+@Tag("integrationTest")
+class GameControllerIntegrationTest {
```
[the remark this answers is not in the document]
Comment (devaajanne) [reply backend/src/test/java/backend/integration/GameControllerIntegrationTests.java]: Fixed

File: backend/build.gradle [hunk 4]
```diff
@@ -31,12 +31,15 @@ dependencies {
 	implementation 'org.springframework.boot:spring-boot-starter-data-jdbc'
 	implementation 'org.springframework.boot:spring-boot-starter-data-jpa'
 	implementation 'org.springframework.boot:spring-boot-starter-validation'
-	implementation 'org.springframework.boot:spring-boot-starter-web'
+	implementation 'org.springframework.boot:spring-boot-starter-webmvc'
```
[the remark this answers is not in the document]
Comment (devaajanne) [reply backend/build.gradle:34]: webmvc is the new dependency in Spring Boot 4+

File: backend/src/test/java/backend/BackendApplicationTests.java [hunk 5]
```diff
@@ -3,12 +3,28 @@
 import org.junit.jupiter.api.Tag;
 import org.junit.jupiter.api.Test;
 import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.boot.test.web.server.LocalServerPort;
```
[the remark this answers is not in the document]
Comment (devaajanne) [reply backend/src/test/java/backend/BackendApplicationTests.java]: Fixed

File: backend/src/test/java/backend/integration/GameControllerIntegrationTests.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (devaajanne) [reply backend/src/test/java/backend/integration/GameControllerIntegrationTests.java]: Fixed

File: backend/src/test/java/backend/integration/GameControllerIntegrationTests.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+  }
+
+  @Test
+  public void getFixedLettersReturns200OkResponseWhenWordLengthIsSeven() throws Exception {
+    mockMvc
+        .perform(get("/api/fixed-letters/FI/7/5").contentType(MediaType.APPLICATION_JSON))
+        .andExpect(status().isOk()) // HTTP 200
+        .andExpect(jsonPath("$.fixedLetters").isArray())
+        .andExpect(jsonPath("$.wordLength").value(7));
+  }
```
[the remark this answers is not in the document]
Comment (devaajanne) [reply backend/src/test/java/backend/integration/GameControllerIntegrationTests.java:74]: Added more tests

File: backend/src/test/java/backend/integration/GameControllerIntegrationTests.java [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+import backend.controller.GameController;
+import org.junit.jupiter.api.Tag;
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.boot.webmvc.test.autoconfigure.AutoConfigureMockMvc;
```
[the remark this answers is not in the document]
Comment (devaajanne) [reply backend/src/test/java/backend/integration/GameControllerIntegrationTests.java:19]: No, the import is correct

File: backend/src/main/java/backend/configuration/GlobalExceptionHandler.java [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+   *
+   * @param exception the IllegalStateException thrown
+   * @return ResponseEntity containing error details and status 204
+   */
+  @ExceptionHandler(IllegalStateException.class)
+  public ResponseEntity<Map<String, String>> handleIllegalStateException(
+      IllegalStateException exception) {
+    Map<String, String> error = new HashMap<>();
+    error.put("error", exception.getMessage());
+    return new ResponseEntity<>(error, HttpStatus.NO_CONTENT);
```
[the remark this answers is not in the document]
Comment (devaajanne) [reply backend/src/main/java/backend/configuration/GlobalExceptionHandler.java]: Fixed
